refinery/core/tests.py

```diff
@@ -611,87 +614,62 @@ def test_verify_deletion_if_workflow_not_used_in_analysis(self):
 
 class DataSetDeletionTest(TestCase):
     """Testing for the deletion of Datasets"""
-
     def setUp(self):
+        self.models_to_be_removed = [
+            Analysis,
+            AnnotatedNode,
+            Assay,
+            DataSet,
+            FileStoreItem,
+            Investigation,
+            InvestigationLink,
+            Node,
+            NodeCollection,
+            Study
+        ]
         self.username = self.password = 'user'
         self.user = User.objects.create_user(
             self.username, '', self.password
         )
-        self.project = Project.objects.create()
-        self.galaxy_instance = GalaxyInstanceFactory()
-        self.isa_archive_file = FileStoreItem.objects.create(
-            datafile=SimpleUploadedFile(
-                'test_file.zip',
-                'Coffee is delicious!')
-        )
-        self.pre_isa_archive_file = FileStoreItem.objects.create(
-            datafile=SimpleUploadedFile(
-                'test_file.txt',
-                'Coffee is delicious!')
-        )
-        self.investigation = Investigation.objects.create(
-                isarchive_file=self.isa_archive_file.uuid,
-                pre_isarchive_file=self.pre_isa_archive_file.uuid
+        self.analyses, self.dataset = \
+            make_analyses_with_single_dataset(
+                1,
+                self.user
             )
 
-        self.workflow_engine = WorkflowEngine.objects.create(
-            instance=self.galaxy_instance
-        )
-        self.workflow = Workflow.objects.create(
-            name="Workflow1", workflow_engine=self.workflow_engine)
-        self.dataset_with_analysis = DataSet.objects.create(
-            name="dataset_with_analysis")
-        self.dataset_without_analysis = \
-            DataSet.objects.create(name="dataset_without_analysis")
-        self.investigation_link = \
-            InvestigationLink.objects.create(
-                investigation=self.investigation,
-                data_set=self.dataset_without_analysis)
+    def _assert_related_objects_exist(self):
+        for model in self.models_to_be_removed:
+            self.assertNotEqual(model.objects.count(), 0)
```

Comment:
What is wrong with having private test case methods that aid in asserting a certain state has been achieved? I've seen this pattern be helpful when making the same assertions many times.

Reply:
I'm not sure I agree with that: The intention of what's being asserted here ("`related_objects_exist`") is more clear than if the two lines inside are copied and pasted where it's used.

Reply:
Yes, and it also makes it difficult to troubleshoot test failures. Two Scoops of Django 22.3.4: "Don't Repeat Yourself Doesn't Apply to Writing Tests"

Reply:
I like two scoops a lot. Sometimes I find the suggestions to be opinionated, and I've learned to use it as a tool for guidance rather than as law. I think in this instance writing:
```
for model in self.models_to_be_removed:
    self.assertNotEqual(model.objects.count(), 0)
```
alone is less clear than calling `self._assert_related_objects_exist()`

Reply:
Tests should be written as simply as possible. Otherwise, troubleshooting failures becomes very difficult which defeats the purpose of writing tests.

Reply:
Looking at this code again I believe this line `self.assertNotEqual(model.objects.count(), 0)` would mask which `model` failed the given assertion:
```
Traceback (most recent call last):
  File "/vagrant/refinery/core/tests.py", line 660, in test_dataset_deletion_removes_related_objects
    self._assert_related_objects_removed()
  File "/vagrant/refinery/core/tests.py", line 646, in _assert_related_objects_removed
    self.assertEqual(model.objects.count(), 0)
AssertionError: 1 != 0
```

Newer Django/Unittest have solved this type of issue with the use of [`subTest()`](https://docs.python.org/3/library/unittest.html#subtests) also see[here](https://www.caktusgroup.com/blog/2017/05/29/subtests-are-best/)

but for now we would have to do something like this for easy debugging of test failures:
```
    def _assert_related_objects_exist(self):
        self.assertGreater(Analysis.objects.count(), 0)
        self.assertGreater(AnnotatedNode.objects.count(), 0)
        self.assertGreater(Assay.objects.count(), 0)
        self.assertGreater(DataSet.objects.count(), 0)
        self.assertGreater(FileStoreItem.objects.count(), 0)
        self.assertGreater(Investigation.objects.count(), 0)
        self.assertGreater(InvestigationLink.objects.count(), 0)
        self.assertGreater(Node.objects.count(), 0)
        self.assertGreater(NodeCollection.objects.count(), 0)
        self.assertGreater(Study.objects.count(), 0)

    def _assert_related_objects_removed(self):
        self.assertEqual(Analysis.objects.count(), 0)
        self.assertEqual(AnnotatedNode.objects.count(), 0)
        self.assertEqual(Assay.objects.count(), 0)
        self.assertEqual(DataSet.objects.count(), 0)
        self.assertEqual(FileStoreItem.objects.count(), 0)
        self.assertEqual(Investigation.objects.count(), 0)
        self.assertEqual(InvestigationLink.objects.count(), 0)
        self.assertEqual(Node.objects.count(), 0)
        self.assertEqual(NodeCollection.objects.count(), 0)
        self.assertEqual(Study.objects.count(), 0)
```

Which would yield something like the following on an error:
```
Failure
Traceback (most recent call last):
  File "/vagrant/refinery/core/tests.py", line 660, in test_dataset_deletion_removes_related_objects
    self._assert_related_objects_removed()
  File "/vagrant/refinery/core/tests.py", line 645, in _assert_related_objects_removed
    self.assertEqual(FileStoreItem.objects.count(), 0)
AssertionError: 1 != 0
```

@hackdna I think that remaining DRY with this approach is favorable. One introducing more tests in this space in the future could easily miss one of the above assertions with a copy/paste approach rather than reusing `_assert_related_objects_exist` or `_assert_related_objects_removed`.

I'm not seeing the difficulty in troubleshooting a failure like the one above.

Reply:
The subTest() method may be useful in limited cases when you have small changes in test conditions, such as running the same test with a range of integers (as it is shown in the unittest docs).

Test cases in the Caktus Group article are too long and should be broken down into individual tests for clarity. Use of subTest() there is superfluous and only obfuscates the code.

Reply:
I think the `subTest()` code would work fine for the problem I illustrated. Allowing for code like this:
```
for model in self.models_to_be_removed:
    with self.subTest(model=model):
        self.assertNotEqual(model.objects.count(), 0)
```

to yield an error like this:

```
Failure (model=FileStoreItem)
Traceback (most recent call last):
  File "/vagrant/refinery/core/tests.py", line 660, in test_dataset_deletion_removes_related_objects
    self._assert_related_objects_removed()
  File "/vagrant/refinery/core/tests.py", line 646, in _assert_related_objects_removed
    self.assertEqual(model.objects.count(), 0)
AssertionError: 1 != 0
```  allowing one to see which `model` had caused the assertion failure.

I linked the Caktus article just to show some usages of `subTest()`. It was my understanding that they chose particularly long tests in the article to better illustrate how it could consolidate testing code.

But thats all besides the fact because I wasn't advocating for the use of this feature since we don't have the ability to right now or in the near future.

---

My point out of my prior comment was that to have easy to debug test failures without succumbing to copy/pasting ~10 assertions every time that we want to check that a dataset's related objects have been removed/remain seems to naturally lead to writing something reusable like: `_assert_related_objects_exist()`.

Reply:
There is no need to check if a data set was created properly in every test. The same goes for deletion. Right now a lot of tests are re-testing functionality covered by other tests. Test code will naturally become DRY when this is no longer the case.

Reply:
I'm confused, we aren't checking "if a data set was created properly in every test". Nor are we for `_assert_related_objects_removed`. Those are only used when they need to be, it just happens to be in  a few places in this PR.

When you say: "Right now a lot of tests are re-testing functionality covered by other tests", which tests are you talking about in the context of this pull request?

I think all of the tests within `DataSetDeletionTest` are distinct use cases that warrant separate testing.

Reply:
For example, the functionality tested by`_assert_related_objects_exist()` is used a total of three times but it does not need to be present in `test_dataset_deletion_removes_related_objects()` and `test_dataset_bulk_deletion_removes_related_objects()` because it has already been tested in `test_transaction_rollback_on_delete_failure()`.

Functionality tested by `_assert_related_objects_removed()` is used a total of two times. Now, if I need to troubleshoot a failure in one of those tests, I'd need to understand not only the logic of the code under test but also, I'd need to look up whatever `_assert_related_objects_removed()` does and then what `self.models_to_be_removed` is. This adds unnecessary complexity and makes troubleshooting more difficult - just to save a few lines of code.